### PR TITLE
Caffe2 hipify python API

### PR DESCRIPTION
**Pull request by rohithkrn**
The goal of this PR is to add hip/miopen options to python APIs and operator tests.

### CODE REVIEW

File: caffe2/python/core.py [hunk 1]
```diff
@@ -2189,27 +2189,37 @@ def extend_ops(self, new_ops):
...
 
-    if src.device_type == CUDA and dst.device_type == CUDA:
-        if src.cuda_gpu_id == dst.cuda_gpu_id:
-            return None
+    if src.device_type == GPU and dst.device_type == GPU:
+        def fun(net, *args, **kw):
+            with DeviceScope(dst):
+                return net.Copy(*args, **kw)
+
+        if workspace.has_gpu_support:
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/python/core.py]: Done.

File: caffe2/python/helpers/conv.py [hunk 2]
```diff
@@ -45,18 +45,22 @@ def _ConvBase(
 
     requested_engine = kwargs.get('engine')
     if requested_engine is not None:
-        if use_cudnn and requested_engine != 'CUDNN':
+        if workspace.has_gpu_support and use_gpu_engine and requested_engine != 'CUDNN':
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/python/helpers/conv.py]: I can't think of a case where it fails.

**Comment by petrex** [comment]
@rohithkrn @ashishfarmer are these the same as our proposed change upstream?

**Comment by rohithkrn** [comment]
@petrex yes, but it's still in WIP. There are some more scripts which needs to be modified.

**Pull request merged**